iroha/src/event.rs

```diff
@@ -0,0 +1,287 @@
+//! Iroha is a quite dynamic system so many events can happen.
+//! This module contains descriptions of such an events and
+//! utilitary Iroha Special Instructions to work with them.
+
+/// Iroha Special Instructions module provides `EventInstruction` enum with all legal types of
+/// events related instructions as variants, implementations of generic Iroha Special Instructions
+/// and the `From/Into` implementations to convert `EventInstruction` variants into generic ISI.
+pub mod isi {
+    use crate::prelude::*;
+    use iroha_derive::*;
+    use parity_scale_codec::{Decode, Encode};
+    use std::time::SystemTime;
+
+    /// Instructions related to different type of Iroha events.
+    /// Some of them are time based triggers, another watch the Blockchain and others
+    /// check the World State View.
+    #[derive(Clone, Debug, Io, Encode, Decode)]
+    pub enum EventInstruction {
+        /// This variant of Iroha Special Instruction will execute instruction when new Block
+        /// will be created.
+        OnBlockCreated(Box<Instruction>),
+        /// This variant of Iroha Special Instruction will execute instruction when Blockchain
+        /// will reach predefined height.
+        OnBlockchainHeight(u64, Box<Instruction>),
+        /// This variant of Iroha Special Instruction will execute instruction when World State
+        /// View change will be detected.
+        OnWorldStateViewChange(Box<Instruction>, Box<Instruction>),
+        /// This variant of Iroha Special Instruction will execute instruction regulary.
+        OnTimestamp(u128, Box<Instruction>),
+    }
+
+    impl EventInstruction {
+        /// Execute `EventInstruction` origin based on the changes in `world_state_view`.
+        pub fn execute(
+            &self,
+            authority: <Account as Identifiable>::Id,
+            world_state_view: &mut WorldStateView,
+        ) -> Result<(), String> {
+            use EventInstruction::*;
+            match self {
+                OnBlockCreated(instruction) => instruction.execute(authority, world_state_view),
+                OnBlockchainHeight(height, instruction) => {
+                    if &world_state_view
+                        .blocks
+                        .last()
+                        .ok_or("Failed to find the last block on the chain.")?
+                        .header
+                        .height
+                        == height
+                    {
+                        instruction.execute(authority, world_state_view)
+                    } else {
+                        Ok(())
+                    }
+                }
+                OnWorldStateViewChange(trigger, instruction) => {
+                    if trigger.execute(authority.clone(), world_state_view).is_ok() {
```

Comment:
To my mind a trigger should not change state, but should be similar to the `Query` but with boolean value.

Reply:
https://jira.hyperledger.org/browse/IR-754?focusedCommentId=69460&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-69460